sklearn/utils/estimator_checks.py

```diff
@@ -2871,3 +2873,28 @@ def check_fit_idempotent(name, estimator_orig):
                 atol=max(tol, 1e-9), rtol=max(tol, 1e-7),
                 err_msg="Idempotency check failed for method {}".format(method)
             )
+
+
+def check_n_features_in(name, estimator_orig):
+    # Make sure that n_features_in_ attribute doesn't exist until fit is
+    # called, and that its value is correct.
+
+    rng = np.random.RandomState(0)
+
+    estimator = clone(estimator_orig)
+    set_random_state(estimator)
+    if 'warm_start' in estimator.get_params().keys():
+        estimator.set_params(warm_start=False)
+
+    n_samples = 100
+    X = rng.normal(loc=100, size=(n_samples, 2))
+    X = _pairwise_estimator_convert_X(X, estimator)
+    if is_regressor(estimator_orig):
+        y = rng.normal(size=n_samples)
+    else:
+        y = rng.randint(low=0, high=2, size=n_samples)
+    y = _enforce_estimator_tags_y(estimator, y)
+
+    assert not hasattr(estimator, 'n_features_in_')
+    estimator.fit(X, y)
+    assert estimator.n_features_in_ == X.shape[1]
```

Comment:
Just to be sure, this will require the new attribute of all estimators in 3rd party packages.

Reply:
Indeed:  maybe we should make this optional for third-party libraries. My proposal would be:

If an estimator sets `n_features_in_` during fit then:

- `n_features_in_` should reflect the number of columns of tabular input fed to fit,
- the contract would be to raise the appropriate `ValueError` when calling `predict` / `transform` / ... with inconsistent inputs.

And in scikit-learn we could have a specific estimators tag in BaseEstimator to force this check by default for all our estimators (apart from aforementioned exceptions: vectorizers...).

Reply:
We can also go through a `FutureWarning` phase in `check_estimator` for third party libs.

Reply:
Indeed, it was decided in the SLEP that this check should raise a warning for the next 2 versions, and only fail after that. 

Thanks for the reminder